backend/app/matrix/background_validator.py

```diff
@@ -0,0 +1,242 @@
+import asyncio
+from typing import Set, Dict, List
+
+import gidgethub.aiohttp
+from nio import ErrorResponse, RoomGetStateEventError, RoomInviteError, RoomKickError
+
+from app.core.app_state import app_state
+from app.core.models import RoomSpecificData, UserData
+from app.github.github_api import GithubAPI
+
+
+class RegisteredUser:
+    def __init__(self, user_data: UserData):
+        self.github_username = user_data.content.github.username
+
+        if self.github_username is not None:
+            gh = gidgethub.aiohttp.GitHubAPI(
+                app_state.http_client.session,
+                requester=self.github_username,
+                oauth_token=user_data.content.github.access_token,
+            )
+
+            self.github_api = GithubAPI(
+                gh=gh,
+                username=self.github_username,
+                default_role=app_state.settings.github.organisation_membership,
+            )
+
+
+class BackgroundValidater:
+    def __init__(self):
+        self.client = app_state.bot_client
+        self.registered_users: Dict[str, RegisteredUser] = dict()
+
+        # List of rooms registered under the application and has "room membership" conditions.
+        self.rooms_with_conditions: List[str] = []
+
+        # Set of rooms where the bot has invite and kick permissions.
+        self.rooms_with_permissions: Set[str] = {}
+        self.background_validator_task = None
+
+    async def create_background_task(self):
+        print("Background validator has started")
+        self.background_validator_task = asyncio.create_task(self.start_task())
+
+    async def cancel_background_task(self):
+        self.background_validator_task.cancel()
+        print("Background validator has stopped")
+
+    async def start_task(self):
+        while True:
+            await self.fetch_rooms_and_users()
+
+            for room_id in self.rooms_with_conditions:
+                # When the bot is either not part of the room or dosen't have the required permissions.
+                if room_id not in self.rooms_with_permissions or room_id not in self.client.rooms:
+                    continue
```

Comment:
Should we be doing some housekeeping here and removing these rooms from the bot state?

Reply:
We will end up deleting the room conditions that a room admin added recently just because the room admin is a bit late in promoting/inviting the bot to the room i.e. add conditions in web app first and then promote bot later.

Reply:
We can instead notify the room admin by sending the message in the room instead?

Reply:
Yeah we wouldn't want that to happen, sending a message sounds good, but not if we are running this loop every few seconds, also sending a message to a space wouldn't really work.

Reply:
Could we set something in the frontend that would show to the admin why the bot isn't working?

Reply:
I guess we can do that.
Maybe I shouldn't let the room admin access the webpage for room conditions until the bot is added and has appropriate permissions for the room?

Reply:
That would be good, to have a holding page there giving instructions on how to add the bot to the room. :+1:

Reply:
For now, I'm removing the rooms and their conditions which don't satisfy any of our two rules (bot not a member, bot doesn't have enough permissions)

It's in commit [fe7b220](https://github.com/MatrixDoctors/matrix-cerberus/pull/56/commits/fe7b2206ba932114798be7ddd41ae2a45c98276c) under PR #56